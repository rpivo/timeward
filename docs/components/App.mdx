---
name: App
menu: Components
---

# App

## Arguments
`App` does not take in any props.

## State

#### isAuthorized: Boolean
`isAuthorized` indicates whether the user is an authorized user. Passing this state down to children will result in specific UI for authorized and non-authorized users.

## Example Usage
`App` is a top-level, single-use component. `isAuthorized` is initialized as `false`. `App` wraps a `Header` component as well as `Route` components, including a special `Route` component called `PrivateRoute` that takes in state `isAuthorized` to determine whether the user can view the route. `App` passes `setIsAuthorized`, the setter for `isAuthorized` down to the `Login` page where the state value can then be set.

```tsx
const App: React.FC = (): JSX.Element => {
  const [isAuthorized, setIsAuthorized] = useState(false);

  return (
    <ThemeProvider theme={Theme}>
      <GlobalStyle />
      <div className='app'>
        <Router>
          <Header isAuthorized={isAuthorized} />
          <Page>
            <Switch>
              <Route
                path="/login"
                render={(): JSX.Element => <Login setIsAuthorized={setIsAuthorized} />}
              />
              <PrivateRoute isAuthorized={isAuthorized} path="*" component={Dashboard} />
            </Switch>
          </Page>
        </Router>
      </div>
    </ThemeProvider>

  );
};
```