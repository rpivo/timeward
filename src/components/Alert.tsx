import React from 'react';
import Alignment from '@components/Alignment';
import StyledAlert from '@styles/components/Alert.styled';

type AlertProps = {
  readonly setShouldDisplayLogoutAlert: (arg: boolean) => void;
  readonly setIsAuthorized: (arg: boolean) => void;
};

const Alert: React.FC<AlertProps> =
  ({ setIsAuthorized, setShouldDisplayLogoutAlert }: AlertProps): JSX.Element => {
    const handleYes = (): void => {
      setShouldDisplayLogoutAlert(false);
      setIsAuthorized(false);
    };

    const handleNo = (): void => setShouldDisplayLogoutAlert(false);

    return (
      <Alignment overlay horizontal vertical>
        <StyledAlert>
          <h2>Are you sure you want to log out?</h2>
          <button onClick={handleYes}>Yes</button>
          <button onClick={handleNo}>No</button>
        </StyledAlert>
      </Alignment>
    );
  };

export default Alert;
