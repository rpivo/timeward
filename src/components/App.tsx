import React from 'react';
import { Dashboard } from '../pages/Dashboard';

export class App extends React.Component {
    render(): JSX.Element { return <Dashboard />; }
}