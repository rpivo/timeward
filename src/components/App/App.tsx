import * as React from 'react';
import { Timer } from '../Timer/Timer';

export class App extends React.Component {
    render(): JSX.Element { return <Timer />; }
}