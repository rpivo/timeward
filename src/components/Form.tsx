import React from 'react';
import { InputProps } from '@components/Input';
import { StyledForm } from '@styles/components/Form.styled';

type FormProps = {
  children: React.ReactNode;
  handleButtonClick: (label: string) => void;
  isSignup: boolean;
  onSubmit: (event: React.FormEvent<HTMLInputElement>) => void;
};

type FormComposition = { Input?: React.FC<InputProps> };

const Form: React.FC<FormProps> & FormComposition =
  ({ children, onSubmit, handleButtonClick, isSignup }: FormProps): JSX.Element =>
    <StyledForm>
      <div className='button-wrapper'>
        <div
          className={`button login ${!isSignup ? 'active' : ''}`}
          onClick={(): void => handleButtonClick('login')}
          role='button'
        >
          LOGIN
        </div>
        <div
          className={`button signup ${isSignup ? 'active' : ''}`}
          onClick={(): void => handleButtonClick('signup')}
          role='button'
        >
          SIGN UP
        </div>
      </div>
      {children}
      <input onClick={onSubmit} type='submit' value={isSignup ? 'SIGN UP' : 'LOGIN'} />
    </StyledForm>;

export default Form;
