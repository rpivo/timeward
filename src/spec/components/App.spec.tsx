import React from 'react';
import renderer from 'react-test-renderer';
import { act } from 'react-dom/test-utils';
import { mount } from 'enzyme';
import CognitoUser from '../__mocks__/amazon-cognito-identity-js/CognitoUser';
import CognitoUserPool from '../__mocks__/amazon-cognito-identity-js/CognitoUserPool';
import AuthenticationDetails from '../__mocks__/amazon-cognito-identity-js/AuthenticationDetails';
import CognitoIdentityCredentials from '../__mocks__/AWS/CognitoIdentityCredentials';
import Alert from '@components/Alert';
import App from '@components/App';
import Form from '@components/Form';

const mockGetIdToken = (): {} => {
  return {
    getJwtToken: jest.fn(),
  };
};

const result = {
  getIdToken: mockGetIdToken,
};

const data = {
  Session: null,
  authenticateUser: jest.fn().mockImplementation((authenticationDetails, callback) => {
    callback.onSuccess(result);
  }),
  authenticationFlowType: 'USER_SRP_AUTH',
  client: {},
  getSignInUserSession: jest.fn().mockReturnValue({}),
  keyPrefix: '',
  pool: {},
  setSignInUserSession: jest.fn(),
  signInUserSession: null,
  storage: {},
  userDataKey: '',
  username: '',
};
jest.mock('amazon-cognito-identity-js', () => {
  return {
    AuthenticationDetails: jest.fn().mockImplementation(
      () => new AuthenticationDetails({})
    ),
    CognitoUser: jest.fn().mockImplementation(() => new CognitoUser(data)),
    CognitoUserPool: jest.fn().mockImplementation(() => new CognitoUserPool({})),
  };
});
jest.mock('aws-sdk/global', () => {
  return {
    CognitoIdentityCredentials: jest.fn().mockImplementation(() =>
      new CognitoIdentityCredentials({})),
    config: {
      credentials: jest.fn().mockImplementation(() =>
        new CognitoIdentityCredentials({})),
      region: 'us-east-1',
    },
  };
});

afterEach(() => {
  jest.restoreAllMocks();
});

describe('App', () => {
  it('should render correctly', () => {
    const tree = renderer
      .create(<App />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  describe('logout', () => {
    it('should log out when the logout button is clicked', () => {
      const event = {} as React.FormEvent<HTMLInputElement>;
      const wrapper = mount(<App />);
      act(() => wrapper.find(Form).props().onSubmit(event));
      wrapper.update();
      wrapper.find('a').at(2).simulate('click');
      const children = wrapper.find(Alert);
      expect(children.length).toEqual(1);
    });
  });
});
