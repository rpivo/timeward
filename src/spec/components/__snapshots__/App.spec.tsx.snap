// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should render correctly 1`] = `
<div
  className="app"
>
  <header
    className="sc-AykKI ipZMMc"
  >
    <h5>
      TIMEWARD
    </h5>
    <nav>
      <a
        href="/dashboard"
        onClick={[Function]}
      >
        Dasbhoard
      </a>
      <a
        href="/login"
        onClick={[Function]}
      >
        Login
      </a>
    </nav>
  </header>
  <main
    className="sc-AykKJ bjPPvP"
  >
    <div
      className="sc-AykKH kyuWRJ"
    >
      <div
        className="sc-AykKC jXbYAx"
      >
        <div
          className="sc-AykKE gPOhkH"
        >
          <span>
            00:00:00
    
          </span>
          <div
            className="sc-AykKD lamqUN"
            onClick={[Function]}
          >
            <svg
              height="50"
              viewBox="0 0 44.398 50"
              width="44.398"
            >
              <path
                d="M28.682,12.087a4.188,4.188,0,0,1,7.254,0L56.741,48.109a4.188,4.188,0,0,1-3.627,6.283H11.5a4.188,4.188,0,0,1-3.627-6.283Z"
                fill="#aaa"
                transform="translate(54.392 -7.309) rotate(90)"
              />
            </svg>
          </div>
          <div
            className="sc-AykKD lamqUN"
            onClick={[Function]}
          >
            <svg
              height="50"
              viewBox="0 0 50 50"
              width="50"
            >
              <path
                d="M4.237,0H45.763A4.237,4.237,0,0,1,50,4.237V45.763A4.237,4.237,0,0,1,45.763,50H4.237A4.237,4.237,0,0,1,0,45.763V4.237A4.237,4.237,0,0,1,4.237,0Z"
                fill="#aaa"
              />
            </svg>
          </div>
        </div>
        <div
          className="sc-AykKF djlmek"
        >
          <div>
            Time Record
          </div>
          <div>
            Time Record
          </div>
          <div>
            Time Record
          </div>
        </div>
      </div>
      <div
        className="sc-AykKG bJtMuy"
      />
      <div
        className="sc-AykKC lmLzZA"
        width="full"
      >
        <div>
          Timeline
        </div>
      </div>
    </div>
  </main>
</div>
`;
